proposals/chunk_polygon.json

```diff
@@ -0,0 +1,80 @@
+{
+    "id": "chunk_polygon",
+    "summary": "Apply a process to spatial chunks of a data cube",
+    "description": "The given data cube is chunked by the given polygons and applies the given process to each individual chunk.",
+    "categories": [
+        "cubes"
+    ],
+    "experimental": true,
+    "parameters": [
+        {
+            "name": "data",
+            "description": "A data cube.",
+            "schema": {
+                "type": "object",
+                "subtype": "raster-cube"
+            }
+        },
+        {
+            "name": "chunks",
+            "description": "A GeoJSON object containing at least one polygon. The provided feature types can be one of the following:\n\n* A `Polygon` or `MultiPolygon` geometry,\n* a `Feature` with a `Polygon` or `MultiPolygon` geometry,\n* a `FeatureCollection` containing at least one `Feature` with `Polygon` or `MultiPolygon` geometries, or\n* a `GeometryCollection` containing `Polygon` or `MultiPolygon` geometries. To maximize interoperability, `GeometryCollection` should be avoided in favour of one of the alternatives above.",
+            "schema": {
+                "type": "object",
+                "subtype": "geojson"
+            }
+        },
+        {
+            "name": "process",
+            "description": "A process that accepts and returns a single data cube and is applied on each individual chunk. The process may consist of multiple sub-processes.",
```

Comment:
> The process may consist of multiple sub-processes.

Isn't this a bit redundant? It's basically the definition of a process.

Reply:
I think there should also be a note that the process should preserve the original dimensions (so no adding/dropping of dimensions, no type changes, ...). 

I'm not sure it should be allowed or forbidden to also change labels, cardinality, resolution,

Reply:
1) Yes for us "pros" that is clear, but we have that in all(?) callback definitions afaik. The reason is that at some point we had people asking whether they could only provide a single process (e.g. "mean") here or more.

2) Good point. I'm not actually sure what the intention here was originally, so forwarding this question to @JeroenVerstraelen and @jdries ...

Reply:
@m-mohr Our implementation assumes that no dimensions can be added/dropped and that the (#rows, #cols) of the input is identical to that of the output. But you can add/drop individual dates or bands.

As it was originally requested in https://github.com/Open-EO/openeo-processes/issues/287, the user needed to be able to view a 4D cube per polygon and create a correlation matrix based on some computation for each date.

I believe that adding/dropping dimensions falls outside of the scope of this process but I think that the preservation of the spatial dimensions can be discussed. Taking into account this documentation for apply https://openeo.org/documentation/1.0/datacubes.html#apply the fact that we call the function 'apply_polygon' could mean:
1. Identical to apply_neighbourhood, apply_kernel: apply a function over every pixel in the datacube using a polygon as a kernel (wrong)
2. Identical to apply_dimension: apply a function over every pixel using the polygon as the input (right)

(In our case we provide a list of polygons so apply_polygons might be an even better name.) So we can use the exact same description as we use for the _apply_ process, it is just that we divided it into an apply for every polygon instead of the whole datacube. The return value of the apply process is:
_A data cube with the newly computed values and the same dimensions. The dimension properties (name, type, labels, reference system and resolution) remain unchanged._

Given that description I think our VITO implementation already permits too much by allowing to add/drop dates or bands.

Reply:
So is it fair to update the description of the returned sub data cube to

> The updated sub data cube with the newly computed values and the same dimensions. The dimension properties (name, type, reference system and resolution) must remain unchanged. The labels can change, but the number of labels must remain unchanged.